# HG changeset patch
# User Rik <rik@octave.org>
# Date 1464757110 25200
#      Tue May 31 21:58:30 2016 -0700
# Node ID 50f49dcae2775c1f15722b5bee54eda2d01faae9
# Parent  45dbd76cae2d18e905e46df8a2df51c7fa1918db
datestr.m: Fix Matlab incompatibility for format #16 (bug #48071).

* datestr.m: When PM indicator is used, normalize hour into the 0-12 range.
Update docstring.  Add BIST test to verify behavior.

diff --git a/scripts/time/datestr.m b/scripts/time/datestr.m
--- a/scripts/time/datestr.m
+++ b/scripts/time/datestr.m
@@ -48,19 +48,19 @@
 ## @item 6 @tab mm/dd                   @tab 09/07
 ## @item 7 @tab dd                      @tab 07
 ## @item 8 @tab ddd                     @tab Thu
 ## @item 9 @tab d                       @tab T
 ## @item 10 @tab yyyy                   @tab 2000
 ## @item 11 @tab yy                     @tab 00
 ## @item 12 @tab mmmyy                  @tab Sep00
 ## @item 13 @tab HH:MM:SS               @tab 15:38:09
-## @item 14 @tab HH:MM:SS PM            @tab 03:38:09 PM
+## @item 14 @tab HH:MM:SS PM            @tab  3:38:09 PM
 ## @item 15 @tab HH:MM                  @tab 15:38
-## @item 16 @tab HH:MM PM               @tab 03:38 PM
+## @item 16 @tab HH:MM PM               @tab  3:38 PM
 ## @item 17 @tab QQ-YY                  @tab Q3-00
 ## @item 18 @tab QQ                     @tab Q3
 ## @item 19 @tab dd/mm                  @tab 07/09
 ## @item 20 @tab dd/mm/yy               @tab 07/09/00
 ## @item 21 @tab mmm.dd,yyyy HH:MM:SS   @tab Sep.07,2000 15:38:08
 ## @item 22 @tab mmm.dd,yyyy            @tab Sep.07,2000
 ## @item 23 @tab mm/dd/yyyy             @tab 09/07/2000
 ## @item 24 @tab dd/mm/yyyy             @tab 07/09/2000
@@ -82,18 +82,18 @@
 ## @item mmmm @tab Full month name                              @tab December
 ## @item mmm  @tab Abbreviated month name                       @tab Dec
 ## @item mm   @tab Numeric month number (padded with zeros)     @tab 01, 08, 12
 ## @item m    @tab First letter of month name (capitalized)     @tab D
 ## @item dddd @tab Full weekday name                            @tab Sunday
 ## @item ddd  @tab Abbreviated weekday name                     @tab Sun
 ## @item dd   @tab Numeric day of month (padded with zeros)     @tab 11
 ## @item d    @tab First letter of weekday name (capitalized)   @tab S
-## @item HH   @tab Hour of day, padded with zeros if PM is set  @tab 09:00
-## @item      @tab and not padded with zeros otherwise          @tab 9:00 AM
+## @item HH   @tab Hour of day, padded with zeros,              @tab 09:00
+## @item      @tab or padded with spaces if PM is set           @tab  9:00 AM
 ## @item MM   @tab Minute of hour (padded with zeros)           @tab 10:05
 ## @item SS   @tab Second of minute (padded with zeros)         @tab 10:05:03
 ## @item FFF  @tab Milliseconds of second (padded with zeros)   @tab 10:05:03.012
 ## @item AM   @tab Use 12-hour time format                      @tab 11:30 AM
 ## @item PM   @tab Use 12-hour time format                      @tab 11:30 PM
 ## @end multitable
 ##
 ## If @var{f} is not specified or is @code{-1}, then use 0, 1 or 16, depending
@@ -215,17 +215,21 @@ function retval = datestr (date, f = [],
 
     df_orig = df;
     df = strrep (df, "AM", "%p");
     df = strrep (df, "PM", "%p");
     if (strcmp (df, df_orig))
       ## PM not set.
       df = strrep (df, "HH", "%H");
     else
-      df = strrep (df, "HH", sprintf ("%2d", v(i,4)));
+      hr = mod (v(i,4), 12);
+      if (hr == 0)
+        hr = 12;
+      endif
+      df = strrep (df, "HH", sprintf ("%2d", hr));
     endif
 
     df = regexprep (df, "[Yy][Yy][Yy][Yy]", "%Y");
 
     df = regexprep (df, "[Yy][Yy]", "%y");
 
     df = regexprep (df, "[Dd][Dd][Dd][Dd]", "%A");
 
@@ -307,24 +311,29 @@ endfunction
 %!assert (datestr (testtime,7), "18")
 %!assert (datestr (testtime,8), "Sun")
 %!assert (datestr (testtime,9), "S")
 %!assert (datestr (testtime,10), "2005")
 %!assert (datestr (testtime,11), "05")
 %!assert (datestr (testtime,12), "Dec05")
 %!assert (datestr (testtime,13), "02:33:17")
 ## Mac OS X interprets %p parameter to strftime as lower case am/pm indicator.
-## Accomodate this, although no other UNIX-based OS does this.
+## Accommodate this, although no other UNIX-based OS does this.
 %!test
 %! obs = toupper (datestr (testtime,14));
 %! assert (obs, " 2:33:17 AM");
 %!assert (datestr (testtime,15), "02:33")
 %!test
 %! obs = toupper (datestr (testtime,16));
 %! assert (obs, " 2:33 AM");
+%!test  # bug #48071
+%! testtime2 = testtime;
+%! testtime2(4) = 15;
+%! obs = toupper (datestr (testtime2,16));
+%! assert (obs, " 3:33 PM");
 %!assert (datestr (testtime,17), "Q4-05")
 %!assert (datestr (testtime,18), "Q4")
 %!assert (datestr (testtime,19), "18/12")
 %!assert (datestr (testtime,20), "18/12/05")
 %!assert (datestr (testtime,21), "Dec.18,2005 02:33:17")
 %!assert (datestr (testtime,22), "Dec.18,2005")
 %!assert (datestr (testtime,23), "12/18/2005")
 %!assert (datestr (testtime,24), "18/12/2005")
